fvm/evm/emulator/state/delta.go

```diff
@@ -0,0 +1,414 @@
+	created        map[gethCommon.Address]interface{}
+	suicided       map[gethCommon.Address]interface{}
+	balances       map[gethCommon.Address]*big.Int
+	nonces         map[gethCommon.Address]uint64
+	codes          map[gethCommon.Address][]byte
+	codeHashes     map[gethCommon.Address]gethCommon.Hash
+
+	// states changes
+	dirtySlots map[types.SlotAddress]interface{}
+	states     map[types.SlotAddress]gethCommon.Hash
+
+	// transient storage
+	transient map[types.SlotAddress]gethCommon.Hash
+
+	// access lists
+	accessListAddresses map[gethCommon.Address]interface{}
+	accessListSlots     map[types.SlotAddress]interface{}
+
+	// logs
+	logs []*gethTypes.Log
+
+	// preimages
+	preimages map[gethCommon.Hash][]byte
+
+	// refund
+	refund uint64
+}
+
+var _ types.HotView = &DeltaView{}
+
+// NewDeltaView constructs a new delta view
+func NewDeltaView(parent types.ReadOnlyView) *DeltaView {
+	return &DeltaView{
+		parent: parent,
+
+		dirtyAddresses:      make(map[gethCommon.Address]interface{}, 0),
+		created:             make(map[gethCommon.Address]interface{}, 0),
+		suicided:            make(map[gethCommon.Address]interface{}, 0),
+		balances:            make(map[gethCommon.Address]*big.Int, 0),
+		nonces:              make(map[gethCommon.Address]uint64, 0),
+		codes:               make(map[gethCommon.Address][]byte, 0),
+		codeHashes:          make(map[gethCommon.Address]gethCommon.Hash, 0),
+		dirtySlots:          make(map[types.SlotAddress]interface{}, 0),
+		states:              make(map[types.SlotAddress]gethCommon.Hash, 0),
+		transient:           make(map[types.SlotAddress]gethCommon.Hash, 0),
+		accessListAddresses: make(map[gethCommon.Address]interface{}, 0),
+		accessListSlots:     make(map[types.SlotAddress]interface{}, 0),
+		logs:                make([]*gethTypes.Log, 0),
+		preimages:           make(map[gethCommon.Hash][]byte, 0),
+
+		// for refund we just copy the data
+		refund: parent.GetRefund(),
+	}
+}
+
+// NewChildView constructs a new delta view having the current view as parent
+func (d *DeltaView) NewChildView() *DeltaView {
+	return NewDeltaView(d)
+}
+
+// Exist returns true if address exists
+//
+// it also returns true for both newly created accounts or accounts that has been flagged for deletion
+func (d *DeltaView) Exist(addr gethCommon.Address) (bool, error) {
+	_, found := d.created[addr]
+	if found {
+		return true, nil
+	}
+	_, found = d.suicided[addr]
+	if found {
+		return true, nil
+	}
+	return d.parent.Exist(addr)
+}
+
+// CreateAccount creates a new account for the given address
+func (d *DeltaView) CreateAccount(addr gethCommon.Address) error {
+	d.created[addr] = struct{}{}
+	// flag the address as dirty
+	d.dirtyAddresses[addr] = struct{}{}
+	return nil
+}
+
+// IsCreated returns true if address has been created in this tx
+func (d *DeltaView) IsCreated(addr gethCommon.Address) bool {
+	_, found := d.created[addr]
+	if found {
+		return true
+	}
+	return d.parent.IsCreated(addr)
```

Comment:
doesn't this recursion lacks termination? I guess it should check if parent is nil?